diff -ur linux-0.11/boot/bootsect.s linux-0.11-deb/boot/bootsect.s
--- linux-0.11/boot/bootsect.s	1991-12-06 06:47:58.000000000 +0800
+++ linux-0.11-deb/boot/bootsect.s	2013-11-21 10:13:19.672825034 +0800
@@ -225,11 +225,11 @@
 	pop ax
 	jmp read_track
 
-/*
- * This procedure turns off the floppy drive motor, so
- * that we enter the kernel in a known state, and
- * don't have to worry about it later.
- */
+
+! This procedure turns off the floppy drive motor, so
+! that we enter the kernel in a known state, and
+! don't have to worry about it later.
+
 kill_motor:
 	push dx
 	mov dx,#0x3f2
diff -ur linux-0.11/boot/head.s linux-0.11-deb/boot/head.s
--- linux-0.11/boot/head.s	1991-11-18 23:05:09.000000000 +0800
+++ linux-0.11-deb/boot/head.s	2013-11-22 10:40:43.219531669 +0800
@@ -12,15 +12,15 @@
  * the page directory.
  */
 .text
-.globl _idt,_gdt,_pg_dir,_tmp_floppy_area
-_pg_dir:
+.globl idt, gdt, pg_dir, tmp_floppy_area, startup_32
+pg_dir:
 startup_32:
 	movl $0x10,%eax
 	mov %ax,%ds
 	mov %ax,%es
 	mov %ax,%fs
 	mov %ax,%gs
-	lss _stack_start,%esp
+	lss stack_start,%esp
 	call setup_idt
 	call setup_gdt
 	movl $0x10,%eax		# reload all the segment registers
@@ -28,7 +28,7 @@
 	mov %ax,%es		# reloaded in 'setup_gdt'
 	mov %ax,%fs
 	mov %ax,%gs
-	lss _stack_start,%esp
+	lss stack_start,%esp
 	xorl %eax,%eax
 1:	incl %eax		# check that A20 really IS enabled
 	movl %eax,0x000000	# loop forever if it isn't
@@ -60,7 +60,7 @@
 	xorl $6,%eax		/* reset MP, set EM */
 	movl %eax,%cr0
 	ret
-.align 2
+.align 4
 1:	.byte 0xDB,0xE4		/* fsetpm for 287, ignored by 387 */
 	ret
 
@@ -81,7 +81,7 @@
 	movw %dx,%ax		/* selector = 0x0008 = cs */
 	movw $0x8E00,%dx	/* interrupt gate - dpl=0, present */
 
-	lea _idt,%edi
+	lea idt,%edi
 	mov $256,%ecx
 rp_sidt:
 	movl %eax,(%edi)
@@ -129,7 +129,7 @@
  * reach to a buffer-block. It needs to be aligned, so that it isn't
  * on a 64kB border.
  */
-_tmp_floppy_area:
+tmp_floppy_area:
 	.fill 1024,1,0
 
 after_page_tables:
@@ -137,7 +137,7 @@
 	pushl $0
 	pushl $0
 	pushl $L6		# return address for main, if it decides to.
-	pushl $_main
+	pushl $main
 	jmp setup_paging
 L6:
 	jmp L6			# main should never return here, but
@@ -146,7 +146,7 @@
 /* This is the default interrupt "handler" :-) */
 int_msg:
 	.asciz "Unknown interrupt\n\r"
-.align 2
+.align 4
 ignore_int:
 	pushl %eax
 	pushl %ecx
@@ -159,7 +159,7 @@
 	mov %ax,%es
 	mov %ax,%fs
 	pushl $int_msg
-	call _printk
+	call printk
 	popl %eax
 	pop %fs
 	pop %es
@@ -194,16 +194,16 @@
  * some kind of marker at them (search for "16Mb"), but I
  * won't guarantee that's all :-( )
  */
-.align 2
+.align 4
 setup_paging:
 	movl $1024*5,%ecx		/* 5 pages - pg_dir+4 page tables */
 	xorl %eax,%eax
 	xorl %edi,%edi			/* pg_dir is at 0x000 */
 	cld;rep;stosl
-	movl $pg0+7,_pg_dir		/* set present bit/user r/w */
-	movl $pg1+7,_pg_dir+4		/*  --------- " " --------- */
-	movl $pg2+7,_pg_dir+8		/*  --------- " " --------- */
-	movl $pg3+7,_pg_dir+12		/*  --------- " " --------- */
+	movl $pg0+7,pg_dir		/* set present bit/user r/w */
+	movl $pg1+7,pg_dir+4		/*  --------- " " --------- */
+	movl $pg2+7,pg_dir+8		/*  --------- " " --------- */
+	movl $pg3+7,pg_dir+12		/*  --------- " " --------- */
 	movl $pg3+4092,%edi
 	movl $0xfff007,%eax		/*  16Mb - 4096 + 7 (r/w user,p) */
 	std
@@ -217,21 +217,21 @@
 	movl %eax,%cr0		/* set paging (PG) bit */
 	ret			/* this also flushes prefetch-queue */
 
-.align 2
+.align 4
 .word 0
 idt_descr:
 	.word 256*8-1		# idt contains 256 entries
-	.long _idt
-.align 2
+	.long idt
+.align 4
 .word 0
 gdt_descr:
 	.word 256*8-1		# so does gdt (not that that's any
-	.long _gdt		# magic number, but it works for me :^)
+	.long gdt		# magic number, but it works for me :^)
 
-	.align 3
-_idt:	.fill 256,8,0		# idt is uninitialized
+	.align 8
+idt:	.fill 256,8,0		# idt is uninitialized
 
-_gdt:	.quad 0x0000000000000000	/* NULL descriptor */
+gdt:	.quad 0x0000000000000000	/* NULL descriptor */
 	.quad 0x00c09a0000000fff	/* 16Mb */
 	.quad 0x00c0920000000fff	/* 16Mb */
 	.quad 0x0000000000000000	/* TEMPORARY - don't use */
diff -ur linux-0.11/fs/bitmap.c linux-0.11-deb/fs/bitmap.c
--- linux-0.11/fs/bitmap.c	1991-11-27 05:31:53.000000000 +0800
+++ linux-0.11-deb/fs/bitmap.c	2013-11-21 17:15:34.717770044 +0800
@@ -14,16 +14,16 @@
 __asm__("cld\n\t" \
 	"rep\n\t" \
 	"stosl" \
-	::"a" (0),"c" (BLOCK_SIZE/4),"D" ((long) (addr)):"cx","di")
+	::"a" (0),"c" (BLOCK_SIZE/4),"D" ((long) (addr)))
 
 #define set_bit(nr,addr) ({\
-register int res __asm__("ax"); \
+register int res; \
 __asm__ __volatile__("btsl %2,%3\n\tsetb %%al": \
 "=a" (res):"0" (0),"r" (nr),"m" (*(addr))); \
 res;})
 
 #define clear_bit(nr,addr) ({\
-register int res __asm__("ax"); \
+register int res; \
 __asm__ __volatile__("btrl %2,%3\n\tsetnb %%al": \
 "=a" (res):"0" (0),"r" (nr),"m" (*(addr))); \
 res;})
@@ -41,7 +41,7 @@
 	"cmpl $8192,%%ecx\n\t" \
 	"jl 1b\n" \
 	"3:" \
-	:"=c" (__res):"c" (0),"S" (addr):"ax","dx","si"); \
+	:"=c" (__res):"c" (0),"S" (addr):"ax","dx"); \
 __res;})
 
 void free_block(int dev, int block)
diff -ur linux-0.11/fs/buffer.c linux-0.11-deb/fs/buffer.c
--- linux-0.11/fs/buffer.c	1991-12-07 04:21:00.000000000 +0800
+++ linux-0.11-deb/fs/buffer.c	2013-11-21 15:35:36.534509882 +0800
@@ -284,8 +284,7 @@
 __asm__("cld\n\t" \
 	"rep\n\t" \
 	"movsl\n\t" \
-	::"c" (BLOCK_SIZE/4),"S" (from),"D" (to) \
-	:"cx","di","si")
+	::"c" (BLOCK_SIZE/4),"S" (from),"D" (to))
 
 /*
  * bread_page reads four buffers into memory at the desired address. It's
diff -ur linux-0.11/fs/exec.c linux-0.11-deb/fs/exec.c
--- linux-0.11/fs/exec.c	1991-12-02 04:01:01.000000000 +0800
+++ linux-0.11-deb/fs/exec.c	2013-11-21 16:54:02.554994999 +0800
@@ -135,10 +135,11 @@
 				offset = p % PAGE_SIZE;
 				if (from_kmem==2)
 					set_fs(old_fs);
-				if (!(pag = (char *) page[p/PAGE_SIZE]) &&
-				    !(pag = (char *) page[p/PAGE_SIZE] =
-				      (unsigned long *) get_free_page())) 
-					return 0;
+				if (!(pag = (char *) page[p/PAGE_SIZE])) {
+					page[p/PAGE_SIZE] = (unsigned long *) get_free_page();
+					if (!(pag = (char *)page[p/PAGE_SIZE]))
+						return 0;
+				}
 				if (from_kmem==2)
 					set_fs(new_fs);
 
diff -ur linux-0.11/fs/Makefile linux-0.11-deb/fs/Makefile
--- linux-0.11/fs/Makefile	1991-12-02 11:21:31.000000000 +0800
+++ linux-0.11-deb/fs/Makefile	2013-11-21 16:24:08.995245117 +0800
@@ -1,9 +1,9 @@
-AR	=gar
-AS	=gas
+AR	=ar
+AS	=as --32
 CC	=gcc
-LD	=gld
-CFLAGS	=-Wall -O -fstrength-reduce -fcombine-regs -fomit-frame-pointer \
-	-mstring-insns -nostdinc -I../include
+LD	=ld -m elf_i386
+CFLAGS	=-m32 -Wall -O -fstrength-reduce  -fomit-frame-pointer \
+	 -nostdinc -I../include -mtune=i386
 CPP	=gcc -E -nostdinc -I../include
 
 .c.s:
diff -ur linux-0.11/fs/namei.c linux-0.11-deb/fs/namei.c
--- linux-0.11/fs/namei.c	1991-11-26 03:19:59.000000000 +0800
+++ linux-0.11-deb/fs/namei.c	2013-11-21 17:12:18.320908035 +0800
@@ -62,7 +62,7 @@
  */
 static int match(int len,const char * name,struct dir_entry * de)
 {
-	register int same __asm__("ax");
+	register int same;
 
 	if (!de || !de->inode || len > NAME_LEN)
 		return 0;
@@ -72,8 +72,7 @@
 		"fs ; repe ; cmpsb\n\t"
 		"setz %%al"
 		:"=a" (same)
-		:"0" (0),"S" ((long) name),"D" ((long) de->name),"c" (len)
-		:"cx","di","si");
+		:"0" (0),"S" ((long) name),"D" ((long) de->name),"c" (len));
 	return same;
 }
 
diff -ur linux-0.11/fs/super.c linux-0.11-deb/fs/super.c
--- linux-0.11/fs/super.c	1991-12-07 04:10:12.000000000 +0800
+++ linux-0.11-deb/fs/super.c	2013-11-21 13:49:06.658471501 +0800
@@ -20,7 +20,7 @@
 
 /* set_bit uses setb, as gas doesn't recognize setc */
 #define set_bit(bitnr,addr) ({ \
-register int __res __asm__("ax"); \
+register int __res; \
 __asm__("bt %2,%3;setb %%al":"=a" (__res):"a" (0),"r" (bitnr),"m" (*(addr))); \
 __res; })
 
diff -ur linux-0.11/include/asm/memory.h linux-0.11-deb/include/asm/memory.h
--- linux-0.11/include/asm/memory.h	1991-06-16 05:54:44.000000000 +0900
+++ linux-0.11-deb/include/asm/memory.h	2013-11-21 17:28:58.632813919 +0800
@@ -8,7 +8,6 @@
 #define memcpy(dest,src,n) ({ \
 void * _res = dest; \
 __asm__ ("cld;rep;movsb" \
-	::"D" ((long)(_res)),"S" ((long)(src)),"c" ((long) (n)) \
-	:"di","si","cx"); \
+         ::"D" ((long)(_res)),"S" ((long)(src)),"c" ((long) (n)));\
 _res; \
 })
diff -ur linux-0.11/include/asm/segment.h linux-0.11-deb/include/asm/segment.h
--- linux-0.11/include/asm/segment.h	1991-11-26 02:48:24.000000000 +0800
+++ linux-0.11-deb/include/asm/segment.h	2013-11-21 16:35:01.633698824 +0800
@@ -24,7 +24,7 @@
 
 extern inline void put_fs_byte(char val,char *addr)
 {
-__asm__ ("movb %0,%%fs:%1"::"r" (val),"m" (*addr));
+__asm__ ("movb %0,%%fs:%1"::"q" (val),"m" (*addr));
 }
 
 extern inline void put_fs_word(short val,short * addr)
diff -ur linux-0.11/include/linux/sched.h linux-0.11-deb/include/linux/sched.h
--- linux-0.11/include/linux/sched.h	1991-11-20 22:40:46.000000000 +0800
+++ linux-0.11-deb/include/linux/sched.h	2013-11-22 09:19:12.120935096 +0800
@@ -170,12 +170,12 @@
  */
 #define switch_to(n) {\
 struct {long a,b;} __tmp; \
-__asm__("cmpl %%ecx,_current\n\t" \
+__asm__("cmpl %%ecx,current\n\t" \
 	"je 1f\n\t" \
 	"movw %%dx,%1\n\t" \
-	"xchgl %%ecx,_current\n\t" \
+	"xchgl %%ecx,current\n\t" \
 	"ljmp %0\n\t" \
-	"cmpl %%ecx,_last_task_used_math\n\t" \
+	"cmpl %%ecx,last_task_used_math\n\t" \
 	"jne 1f\n\t" \
 	"clts\n" \
 	"1:" \
@@ -193,8 +193,7 @@
 	::"m" (*((addr)+2)), \
 	  "m" (*((addr)+4)), \
 	  "m" (*((addr)+7)), \
-	  "d" (base) \
-	:"dx")
+	  "d" (base))
 
 #define _set_limit(addr,limit) \
 __asm__("movw %%dx,%0\n\t" \
@@ -205,8 +204,7 @@
 	"movb %%dl,%1" \
 	::"m" (*(addr)), \
 	  "m" (*((addr)+6)), \
-	  "d" (limit) \
-	:"dx")
+	  "d" (limit))
 
 #define set_base(ldt,base) _set_base( ((char *)&(ldt)) , base )
 #define set_limit(ldt,limit) _set_limit( ((char *)&(ldt)) , (limit-1)>>12 )
diff -ur linux-0.11/include/string.h linux-0.11-deb/include/string.h
--- linux-0.11/include/string.h	1991-09-17 23:04:09.000000000 +0800
+++ linux-0.11-deb/include/string.h	2013-11-21 17:50:16.593003793 +0800
@@ -31,7 +31,7 @@
 	"stosb\n\t"
 	"testb %%al,%%al\n\t"
 	"jne 1b"
-	::"S" (src),"D" (dest):"si","di","ax");
+	::"S" (src),"D" (dest):"ax");
 return dest;
 }
 
@@ -47,7 +47,7 @@
 	"rep\n\t"
 	"stosb\n"
 	"2:"
-	::"S" (src),"D" (dest),"c" (count):"si","di","ax","cx");
+	::"S" (src),"D" (dest),"c" (count):"ax");
 return dest;
 }
 
@@ -61,7 +61,7 @@
 	"stosb\n\t"
 	"testb %%al,%%al\n\t"
 	"jne 1b"
-	::"S" (src),"D" (dest),"a" (0),"c" (0xffffffff):"si","di","ax","cx");
+	::"S" (src),"D" (dest),"a" (0),"c" (0xffffffff));
 return dest;
 }
 
@@ -80,8 +80,7 @@
 	"jne 1b\n"
 	"2:\txorl %2,%2\n\t"
 	"stosb"
-	::"S" (src),"D" (dest),"a" (0),"c" (0xffffffff),"g" (count)
-	:"si","di","ax","cx");
+	::"S" (src),"D" (dest),"a" (0),"c" (0xffffffff),"g" (count));
 return dest;
 }
 
@@ -100,7 +99,7 @@
 	"jl 3f\n\t"
 	"negl %%eax\n"
 	"3:"
-	:"=a" (__res):"D" (cs),"S" (ct):"si","di");
+	:"=a" (__res):"D" (cs),"S" (ct));
 return __res;
 }
 
@@ -121,7 +120,7 @@
 	"jl 4f\n\t"
 	"negl %%eax\n"
 	"4:"
-	:"=a" (__res):"D" (cs),"S" (ct),"c" (count):"si","di","cx");
+	:"=a" (__res):"D" (cs),"S" (ct),"c" (count));
 return __res;
 }
 
@@ -138,7 +137,7 @@
 	"movl $1,%1\n"
 	"2:\tmovl %1,%0\n\t"
 	"decl %0"
-	:"=a" (__res):"S" (s),"0" (c):"si");
+	:"=a" (__res):"S" (s),"0" (c));
 return __res;
 }
 
@@ -154,7 +153,7 @@
 	"decl %0\n"
 	"2:\ttestb %%al,%%al\n\t"
 	"jne 1b"
-	:"=d" (__res):"0" (0),"S" (s),"a" (c):"ax","si");
+	:"=d" (__res):"0" (0),"S" (s),"a" (c));
 return __res;
 }
 
@@ -178,7 +177,7 @@
 	"je 1b\n"
 	"2:\tdecl %0"
 	:"=S" (__res):"a" (0),"c" (0xffffffff),"0" (cs),"g" (ct)
-	:"ax","cx","dx","di");
+	:"dx","di");
 return __res-cs;
 }
 
@@ -202,7 +201,7 @@
 	"jne 1b\n"
 	"2:\tdecl %0"
 	:"=S" (__res):"a" (0),"c" (0xffffffff),"0" (cs),"g" (ct)
-	:"ax","cx","dx","di");
+	:"dx","di");
 return __res-cs;
 }
 
@@ -229,7 +228,7 @@
 	"2:\txorl %0,%0\n"
 	"3:"
 	:"=S" (__res):"a" (0),"c" (0xffffffff),"0" (cs),"g" (ct)
-	:"ax","cx","dx","di");
+	:"dx","di");
 return __res;
 }
 
@@ -256,7 +255,7 @@
 	"xorl %%eax,%%eax\n\t"
 	"2:"
 	:"=a" (__res):"0" (0),"c" (0xffffffff),"S" (cs),"g" (ct)
-	:"cx","dx","di","si");
+	:"dx","di");
 return __res;
 }
 
@@ -268,7 +267,7 @@
 	"scasb\n\t"
 	"notl %0\n\t"
 	"decl %0"
-	:"=c" (__res):"D" (s),"a" (0),"0" (0xffffffff):"di");
+	:"=c" (__res):"D" (s),"a" (0),"0" (0xffffffff));
 return __res;
 }
 
@@ -338,8 +337,7 @@
 __asm__("cld\n\t"
 	"rep\n\t"
 	"movsb"
-	::"c" (n),"S" (src),"D" (dest)
-	:"cx","si","di");
+	::"c" (n),"S" (src),"D" (dest));
 return dest;
 }
 
@@ -349,14 +347,12 @@
 __asm__("cld\n\t"
 	"rep\n\t"
 	"movsb"
-	::"c" (n),"S" (src),"D" (dest)
-	:"cx","si","di");
+	::"c" (n),"S" (src),"D" (dest));
 else
 __asm__("std\n\t"
 	"rep\n\t"
 	"movsb"
-	::"c" (n),"S" (src+n-1),"D" (dest+n-1)
-	:"cx","si","di");
+	::"c" (n),"S" (src+n-1),"D" (dest+n-1));
 return dest;
 }
 
@@ -371,8 +367,7 @@
 	"jl 1f\n\t"
 	"negl %%eax\n"
 	"1:"
-	:"=a" (__res):"0" (0),"D" (cs),"S" (ct),"c" (count)
-	:"si","di","cx");
+	:"=a" (__res):"0" (0),"D" (cs),"S" (ct),"c" (count));
 return __res;
 }
 
@@ -387,8 +382,7 @@
 	"je 1f\n\t"
 	"movl $1,%0\n"
 	"1:\tdecl %0"
-	:"=D" (__res):"a" (c),"D" (cs),"c" (count)
-	:"cx");
+	:"=D" (__res):"a" (c),"D" (cs),"c" (count));
 return __res;
 }
 
@@ -397,8 +391,7 @@
 __asm__("cld\n\t"
 	"rep\n\t"
 	"stosb"
-	::"a" (c),"D" (s),"c" (count)
-	:"cx","di");
+	::"a" (c),"D" (s),"c" (count));
 return s;
 }
 
diff -ur linux-0.11/include/unistd.h linux-0.11-deb/include/unistd.h
--- linux-0.11/include/unistd.h	1991-11-26 04:18:55.000000000 +0800
+++ linux-0.11-deb/include/unistd.h	2013-11-21 13:31:44.952409659 +0800
@@ -207,7 +207,12 @@
 volatile void exit(int status);
 volatile void _exit(int status);
 int fcntl(int fildes, int cmd, ...);
+/*int fork(void);*/
+#ifndef __IN_MAIN__
 int fork(void);
+int pause(void);
+int sync(void);
+#endif
 int getpid(void);
 int getuid(void);
 int geteuid(void);
@@ -221,7 +226,7 @@
 int mount(const char * specialfile, const char * dir, int rwflag);
 int nice(int val);
 int open(const char * filename, int flag, ...);
-int pause(void);
+/*int pause(void);*/
 int pipe(int * fildes);
 int read(int fildes, char * buf, off_t count);
 int setpgrp(void);
@@ -232,7 +237,7 @@
 int stat(const char * filename, struct stat * stat_buf);
 int fstat(int fildes, struct stat * stat_buf);
 int stime(time_t * tptr);
-int sync(void);
+/*int sync(void);*/
 time_t time(time_t * tloc);
 time_t times(struct tms * tbuf);
 int ulimit(int cmd, long limit);
diff -ur linux-0.11/init/main.c linux-0.11-deb/init/main.c
--- linux-0.11/init/main.c	1991-12-08 03:55:21.000000000 +0800
+++ linux-0.11-deb/init/main.c	2013-11-21 13:28:36.973586190 +0800
@@ -4,6 +4,7 @@
  *  (C) 1991  Linus Torvalds
  */
 
+#define __IN_MAIN__
 #define __LIBRARY__
 #include <unistd.h>
 #include <time.h>
diff -ur linux-0.11/kernel/asm.s linux-0.11-deb/kernel/asm.s
--- linux-0.11/kernel/asm.s	1991-11-18 08:30:28.000000000 +0800
+++ linux-0.11-deb/kernel/asm.s	2013-11-22 10:18:45.198321692 +0800
@@ -11,13 +11,13 @@
  * the fpu must be properly saved/resored. This hasn't been tested.
  */
 
-.globl _divide_error,_debug,_nmi,_int3,_overflow,_bounds,_invalid_op
-.globl _double_fault,_coprocessor_segment_overrun
-.globl _invalid_TSS,_segment_not_present,_stack_segment
-.globl _general_protection,_coprocessor_error,_irq13,_reserved
+.globl divide_error,debug,nmi,int3,overflow,bounds,invalid_op
+.globl double_fault,coprocessor_segment_overrun
+.globl invalid_TSS,segment_not_present,stack_segment
+.globl general_protection,coprocessor_error,irq13,reserved
 
-_divide_error:
-	pushl $_do_divide_error
+divide_error:
+	pushl $do_divide_error
 no_error_code:
 	xchgl %eax,(%esp)
 	pushl %ebx
@@ -50,39 +50,39 @@
 	popl %eax
 	iret
 
-_debug:
-	pushl $_do_int3		# _do_debug
+debug:
+	pushl $do_int3		# do_debug
 	jmp no_error_code
 
-_nmi:
-	pushl $_do_nmi
+nmi:
+	pushl $do_nmi
 	jmp no_error_code
 
-_int3:
-	pushl $_do_int3
+int3:
+	pushl $do_int3
 	jmp no_error_code
 
-_overflow:
-	pushl $_do_overflow
+overflow:
+	pushl $do_overflow
 	jmp no_error_code
 
-_bounds:
-	pushl $_do_bounds
+bounds:
+	pushl $do_bounds
 	jmp no_error_code
 
-_invalid_op:
-	pushl $_do_invalid_op
+invalid_op:
+	pushl $do_invalid_op
 	jmp no_error_code
 
-_coprocessor_segment_overrun:
-	pushl $_do_coprocessor_segment_overrun
+coprocessor_segment_overrun:
+	pushl $do_coprocessor_segment_overrun
 	jmp no_error_code
 
-_reserved:
-	pushl $_do_reserved
+reserved:
+	pushl $do_reserved
 	jmp no_error_code
 
-_irq13:
+irq13:
 	pushl %eax
 	xorb %al,%al
 	outb %al,$0xF0
@@ -92,10 +92,10 @@
 1:	jmp 1f
 1:	outb %al,$0xA0
 	popl %eax
-	jmp _coprocessor_error
+	jmp coprocessor_error
 
-_double_fault:
-	pushl $_do_double_fault
+double_fault:
+	pushl $do_double_fault
 error_code:
 	xchgl %eax,4(%esp)		# error code <-> %eax
 	xchgl %ebx,(%esp)		# &function <-> %ebx
@@ -128,19 +128,19 @@
 	popl %eax
 	iret
 
-_invalid_TSS:
-	pushl $_do_invalid_TSS
+invalid_TSS:
+	pushl $do_invalid_TSS
 	jmp error_code
 
-_segment_not_present:
-	pushl $_do_segment_not_present
+segment_not_present:
+	pushl $do_segment_not_present
 	jmp error_code
 
-_stack_segment:
-	pushl $_do_stack_segment
+stack_segment:
+	pushl $do_stack_segment
 	jmp error_code
 
-_general_protection:
-	pushl $_do_general_protection
+general_protection:
+	pushl $do_general_protection
 	jmp error_code
 
diff -ur linux-0.11/kernel/blk_drv/blk.h linux-0.11-deb/kernel/blk_drv/blk.h
--- linux-0.11/kernel/blk_drv/blk.h	1991-12-06 03:58:01.000000000 +0800
+++ linux-0.11-deb/kernel/blk_drv/blk.h	2013-11-21 17:21:18.775163964 +0800
@@ -84,7 +84,7 @@
 #define DEVICE_ON(device)
 #define DEVICE_OFF(device)
 
-#elif
+#else
 /* unknown blk device */
 #error "unknown blk device"
 
diff -ur linux-0.11/kernel/blk_drv/floppy.c linux-0.11-deb/kernel/blk_drv/floppy.c
--- linux-0.11/kernel/blk_drv/floppy.c	1991-12-07 08:00:38.000000000 +0800
+++ linux-0.11-deb/kernel/blk_drv/floppy.c	2013-11-21 17:24:17.399832602 +0800
@@ -154,8 +154,7 @@
 
 #define copy_buffer(from,to) \
 __asm__("cld ; rep ; movsl" \
-	::"c" (BLOCK_SIZE/4),"S" ((long)(from)),"D" ((long)(to)) \
-	:"cx","di","si")
+	::"c" (BLOCK_SIZE/4),"S" ((long)(from)),"D" ((long)(to)))
 
 static void setup_DMA(void)
 {
diff -ur linux-0.11/kernel/blk_drv/hd.c linux-0.11-deb/kernel/blk_drv/hd.c
--- linux-0.11/kernel/blk_drv/hd.c	1991-12-06 03:58:17.000000000 +0800
+++ linux-0.11-deb/kernel/blk_drv/hd.c	2013-11-21 17:26:45.804361033 +0800
@@ -59,10 +59,10 @@
 } hd[5*MAX_HD]={{0,0},};
 
 #define port_read(port,buf,nr) \
-__asm__("cld;rep;insw"::"d" (port),"D" (buf),"c" (nr):"cx","di")
+__asm__("cld;rep;insw"::"d" (port),"D" (buf),"c" (nr))
 
 #define port_write(port,buf,nr) \
-__asm__("cld;rep;outsw"::"d" (port),"S" (buf),"c" (nr):"cx","si")
+__asm__("cld;rep;outsw"::"d" (port),"S" (buf),"c" (nr))
 
 extern void hd_interrupt(void);
 extern void rd_load(void);
diff -ur linux-0.11/kernel/blk_drv/Makefile linux-0.11-deb/kernel/blk_drv/Makefile
--- linux-0.11/kernel/blk_drv/Makefile	1991-12-06 03:59:42.000000000 +0800
+++ linux-0.11-deb/kernel/blk_drv/Makefile	2013-11-21 16:24:08.991245134 +0800
@@ -6,13 +6,13 @@
 # unless it's something special (ie not a .c file).
 #
 
-AR	=gar
-AS	=gas
-LD	=gld
+AR	=ar
+AS	=as --32
+LD	=ld -m elf_i386
 LDFLAGS	=-s -x
 CC	=gcc
-CFLAGS	=-Wall -O -fstrength-reduce -fomit-frame-pointer -fcombine-regs \
-	-finline-functions -mstring-insns -nostdinc -I../../include
+CFLAGS	=-m32 -Wall -O -fstrength-reduce -fomit-frame-pointer  \
+	-finline-functions  -nostdinc -I../../include -mtune=i386
 CPP	=gcc -E -nostdinc -I../../include
 
 .c.s:
diff -ur linux-0.11/kernel/chr_drv/console.c linux-0.11-deb/kernel/chr_drv/console.c
--- linux-0.11/kernel/chr_drv/console.c	1991-11-24 02:41:21.000000000 +0800
+++ linux-0.11-deb/kernel/chr_drv/console.c	2013-11-22 09:55:29.652586785 +0800
@@ -116,14 +116,14 @@
 				__asm__("cld\n\t"
 					"rep\n\t"
 					"movsl\n\t"
-					"movl _video_num_columns,%1\n\t"
+					"movl video_num_columns,%1\n\t"
 					"rep\n\t"
 					"stosw"
 					::"a" (video_erase_char),
 					"c" ((video_num_lines-1)*video_num_columns>>1),
 					"D" (video_mem_start),
-					"S" (origin)
-					:"cx","di","si");
+					"S" (origin));
+					
 				scr_end -= origin-video_mem_start;
 				pos -= origin-video_mem_start;
 				origin = video_mem_start;
@@ -133,22 +133,22 @@
 					"stosw"
 					::"a" (video_erase_char),
 					"c" (video_num_columns),
-					"D" (scr_end-video_size_row)
-					:"cx","di");
+					"D" (scr_end-video_size_row));
+
 			}
 			set_origin();
 		} else {
 			__asm__("cld\n\t"
 				"rep\n\t"
 				"movsl\n\t"
-				"movl _video_num_columns,%%ecx\n\t"
+				"movl video_num_columns,%%ecx\n\t"
 				"rep\n\t"
 				"stosw"
 				::"a" (video_erase_char),
 				"c" ((bottom-top-1)*video_num_columns>>1),
 				"D" (origin+video_size_row*top),
-				"S" (origin+video_size_row*(top+1))
-				:"cx","di","si");
+				"S" (origin+video_size_row*(top+1)));
+
 		}
 	}
 	else		/* Not EGA/VGA */
@@ -156,14 +156,14 @@
 		__asm__("cld\n\t"
 			"rep\n\t"
 			"movsl\n\t"
-			"movl _video_num_columns,%%ecx\n\t"
+			"movl video_num_columns,%%ecx\n\t"
 			"rep\n\t"
 			"stosw"
 			::"a" (video_erase_char),
 			"c" ((bottom-top-1)*video_num_columns>>1),
 			"D" (origin+video_size_row*top),
-			"S" (origin+video_size_row*(top+1))
-			:"cx","di","si");
+			"S" (origin+video_size_row*(top+1)));
+
 	}
 }
 
@@ -175,14 +175,14 @@
 			"rep\n\t"
 			"movsl\n\t"
 			"addl $2,%%edi\n\t"	/* %edi has been decremented by 4 */
-			"movl _video_num_columns,%%ecx\n\t"
+			"movl video_num_columns,%%ecx\n\t"
 			"rep\n\t"
 			"stosw"
 			::"a" (video_erase_char),
 			"c" ((bottom-top-1)*video_num_columns>>1),
 			"D" (origin+video_size_row*bottom-4),
-			"S" (origin+video_size_row*(bottom-1)-4)
-			:"ax","cx","di","si");
+			"S" (origin+video_size_row*(bottom-1)-4));
+
 	}
 	else		/* Not EGA/VGA */
 	{
@@ -190,14 +190,14 @@
 			"rep\n\t"
 			"movsl\n\t"
 			"addl $2,%%edi\n\t"	/* %edi has been decremented by 4 */
-			"movl _video_num_columns,%%ecx\n\t"
+			"movl video_num_columns,%%ecx\n\t"
 			"rep\n\t"
 			"stosw"
 			::"a" (video_erase_char),
 			"c" ((bottom-top-1)*video_num_columns>>1),
 			"D" (origin+video_size_row*bottom-4),
-			"S" (origin+video_size_row*(bottom-1)-4)
-			:"ax","cx","di","si");
+			"S" (origin+video_size_row*(bottom-1)-4));
+
 	}
 }
 
@@ -261,8 +261,8 @@
 		"rep\n\t"
 		"stosw\n\t"
 		::"c" (count),
-		"D" (start),"a" (video_erase_char)
-		:"cx","di");
+		"D" (start),"a" (video_erase_char));
+		
 }
 
 static void csi_K(int par)
@@ -292,8 +292,8 @@
 		"rep\n\t"
 		"stosw\n\t"
 		::"c" (count),
-		"D" (start),"a" (video_erase_char)
-		:"cx","di");
+		"D" (start),"a" (video_erase_char));
+		
 }
 
 void csi_m(void)
@@ -458,10 +458,10 @@
 						pos -= video_size_row;
 						lf();
 					}
-					__asm__("movb _attr,%%ah\n\t"
+					__asm__("movb attr,%%ah\n\t"
 						"movw %%ax,%1\n\t"
-						::"a" (c),"m" (*(short *)pos)
-						:"ax");
+						::"a" (c),"m" (*(short *)pos));
+						
 					pos += 2;
 					x++;
 				} else if (c==27)
diff -ur linux-0.11/kernel/chr_drv/keyboard.S linux-0.11-deb/kernel/chr_drv/keyboard.S
--- linux-0.11/kernel/chr_drv/keyboard.S	1991-12-04 23:07:58.000000000 +0800
+++ linux-0.11-deb/kernel/chr_drv/keyboard.S	2013-11-22 10:19:31.354385591 +0800
@@ -13,7 +13,7 @@
 #include <linux/config.h>
 
 .text
-.globl _keyboard_interrupt
+.globl keyboard_interrupt
 
 /*
  * these are for the keyboard read functions
@@ -34,7 +34,7 @@
  *  keyboard scan-code and converts it into the appropriate
  *  ascii character(s).
  */
-_keyboard_interrupt:
+keyboard_interrupt:
 	pushl %eax
 	pushl %ebx
 	pushl %ecx
@@ -44,7 +44,7 @@
 	movl $0x10,%eax
 	mov %ax,%ds
 	mov %ax,%es
-	xorl %al,%al		/* %eax is scan code */
+	xor %al,%al		/* %eax is scan code */
 	inb $0x60,%al
 	cmpb $0xe0,%al
 	je set_e0
@@ -66,7 +66,7 @@
 	movb $0x20,%al
 	outb %al,$0x20
 	pushl $0
-	call _do_tty_interrupt
+	call do_tty_interrupt
 	addl $4,%esp
 	pop %es
 	pop %ds
@@ -88,7 +88,7 @@
 put_queue:
 	pushl %ecx
 	pushl %edx
-	movl _table_list,%edx		# read-queue for console
+	movl table_list,%edx		# read-queue for console
 	movl head(%edx),%ecx
 1:	movb %al,buf(%edx,%ecx)
 	incl %ecx
@@ -211,7 +211,7 @@
 	pushl %eax
 	pushl %ecx
 	pushl %edx
-	call _show_stat
+	call show_stat
 	popl %edx
 	popl %ecx
 	popl %eax
diff -ur linux-0.11/kernel/chr_drv/Makefile linux-0.11-deb/kernel/chr_drv/Makefile
--- linux-0.11/kernel/chr_drv/Makefile	1991-12-02 11:21:41.000000000 +0800
+++ linux-0.11-deb/kernel/chr_drv/Makefile	2013-11-21 16:24:08.991245134 +0800
@@ -6,13 +6,13 @@
 # unless it's something special (ie not a .c file).
 #
 
-AR	=gar
-AS	=gas
-LD	=gld
+AR	=ar
+AS	=as --32
+LD	=ld -m elf_i386
 LDFLAGS	=-s -x
 CC	=gcc
-CFLAGS	=-Wall -O -fstrength-reduce -fomit-frame-pointer -fcombine-regs \
-	-finline-functions -mstring-insns -nostdinc -I../../include
+CFLAGS	=-m32 -Wall -O -fstrength-reduce -fomit-frame-pointer  \
+	-finline-functions  -nostdinc -I../../include -mtune=i386
 CPP	=gcc -E -nostdinc -I../../include
 
 .c.s:
diff -ur linux-0.11/kernel/chr_drv/rs_io.s linux-0.11-deb/kernel/chr_drv/rs_io.s
--- linux-0.11/kernel/chr_drv/rs_io.s	1991-10-02 22:16:30.000000000 +0800
+++ linux-0.11-deb/kernel/chr_drv/rs_io.s	2013-11-22 10:01:02.572961400 +0800
@@ -11,7 +11,7 @@
  */
 
 .text
-.globl _rs1_interrupt,_rs2_interrupt
+.globl rs1_interrupt, rs2_interrupt
 
 size	= 1024				/* must be power of two !
 					   and must match the value
@@ -30,13 +30,13 @@
  * These are the actual interrupt routines. They look where
  * the interrupt is coming from, and take appropriate action.
  */
-.align 2
-_rs1_interrupt:
-	pushl $_table_list+8
+.align 4
+rs1_interrupt:
+	pushl $table_list+8
 	jmp rs_int
-.align 2
-_rs2_interrupt:
-	pushl $_table_list+16
+.align 4
+rs2_interrupt:
+	pushl $table_list+16
 rs_int:
 	pushl %edx
 	pushl %ecx
@@ -79,23 +79,23 @@
 jmp_table:
 	.long modem_status,write_char,read_char,line_status
 
-.align 2
+.align 4
 modem_status:
 	addl $6,%edx		/* clear intr by reading modem status reg */
 	inb %dx,%al
 	ret
 
-.align 2
+.align 4
 line_status:
 	addl $5,%edx		/* clear intr by reading line status reg. */
 	inb %dx,%al
 	ret
 
-.align 2
+.align 4
 read_char:
 	inb %dx,%al
 	movl %ecx,%edx
-	subl $_table_list,%edx
+	subl $table_list,%edx
 	shrl $3,%edx
 	movl (%ecx),%ecx		# read-queue
 	movl head(%ecx),%ebx
@@ -106,11 +106,11 @@
 	je 1f
 	movl %ebx,head(%ecx)
 1:	pushl %edx
-	call _do_tty_interrupt
+	call do_tty_interrupt
 	addl $4,%esp
 	ret
 
-.align 2
+.align 4
 write_char:
 	movl 4(%ecx),%ecx		# write-queue
 	movl head(%ecx),%ebx
@@ -132,7 +132,7 @@
 	cmpl head(%ecx),%ebx
 	je write_buffer_empty
 	ret
-.align 2
+.align 4
 write_buffer_empty:
 	movl proc_list(%ecx),%ebx	# wake up sleeping process
 	testl %ebx,%ebx			# is there any?
diff -ur linux-0.11/kernel/Makefile linux-0.11-deb/kernel/Makefile
--- linux-0.11/kernel/Makefile	1991-12-02 11:21:37.000000000 +0800
+++ linux-0.11-deb/kernel/Makefile	2013-11-21 16:24:08.991245134 +0800
@@ -6,13 +6,13 @@
 # unless it's something special (ie not a .c file).
 #
 
-AR	=gar
-AS	=gas
-LD	=gld
+AR	=ar
+AS	=as --32
+LD	=ld -m elf_i386
 LDFLAGS	=-s -x
 CC	=gcc
-CFLAGS	=-Wall -O -fstrength-reduce -fomit-frame-pointer -fcombine-regs \
-	-finline-functions -mstring-insns -nostdinc -I../include
+CFLAGS	=-m32 -Wall -O -fstrength-reduce -fomit-frame-pointer  \
+	-finline-functions  -nostdinc -I../include -mtune=i386
 CPP	=gcc -E -nostdinc -I../include
 
 .c.s:
diff -ur linux-0.11/kernel/math/Makefile linux-0.11-deb/kernel/math/Makefile
--- linux-0.11/kernel/math/Makefile	1991-11-18 08:21:45.000000000 +0800
+++ linux-0.11-deb/kernel/math/Makefile	2013-11-21 16:24:08.991245134 +0800
@@ -6,13 +6,13 @@
 # unless it's something special (ie not a .c file).
 #
 
-AR	=gar
-AS	=gas
-LD	=gld
+AR	=ar
+AS	=as --32
+LD	=ld -m elf_i386
 LDFLAGS	=-s -x
 CC	=gcc
-CFLAGS	=-Wall -O -fstrength-reduce -fomit-frame-pointer -fcombine-regs \
-	-finline-functions -mstring-insns -nostdinc -I../../include
+CFLAGS	=-m32 -Wall -O -fstrength-reduce -fomit-frame-pointer  \
+	-finline-functions  -nostdinc -I../../include -mtune=i386
 CPP	=gcc -E -nostdinc -I../../include
 
 .c.s:
diff -ur linux-0.11/kernel/printk.c linux-0.11-deb/kernel/printk.c
--- linux-0.11/kernel/printk.c	1991-10-02 22:16:29.000000000 +0800
+++ linux-0.11-deb/kernel/printk.c	2013-11-22 10:16:59.946175486 +0800
@@ -30,9 +30,9 @@
 		"push %%ds\n\t"
 		"pop %%fs\n\t"
 		"pushl %0\n\t"
-		"pushl $_buf\n\t"
+		"pushl $buf\n\t"
 		"pushl $0\n\t"
-		"call _tty_write\n\t"
+		"call tty_write\n\t"
 		"addl $8,%%esp\n\t"
 		"popl %0\n\t"
 		"pop %%fs"
diff -ur linux-0.11/kernel/system_call.s linux-0.11-deb/kernel/system_call.s
--- linux-0.11/kernel/system_call.s	1991-12-04 21:56:34.000000000 +0800
+++ linux-0.11-deb/kernel/system_call.s	2013-11-22 09:32:36.041508638 +0800
@@ -64,20 +64,20 @@
  * Ok, I get parallel printer interrupts while using the floppy for some
  * strange reason. Urgel. Now I just ignore them.
  */
-.globl _system_call,_sys_fork,_timer_interrupt,_sys_execve
-.globl _hd_interrupt,_floppy_interrupt,_parallel_interrupt
-.globl _device_not_available, _coprocessor_error
+.globl system_call, sys_fork, timer_interrupt, sys_execve
+.globl hd_interrupt, floppy_interrupt, parallel_interrupt
+.globl device_not_available, coprocessor_error
 
-.align 2
+.align 4
 bad_sys_call:
 	movl $-1,%eax
 	iret
-.align 2
+.align 4
 reschedule:
 	pushl $ret_from_sys_call
-	jmp _schedule
-.align 2
-_system_call:
+	jmp schedule
+.align 4
+system_call:
 	cmpl $nr_system_calls-1,%eax
 	ja bad_sys_call
 	push %ds
@@ -91,16 +91,16 @@
 	mov %dx,%es
 	movl $0x17,%edx		# fs points to local data space
 	mov %dx,%fs
-	call _sys_call_table(,%eax,4)
+	call sys_call_table(,%eax,4)
 	pushl %eax
-	movl _current,%eax
+	movl current,%eax
 	cmpl $0,state(%eax)		# state
 	jne reschedule
 	cmpl $0,counter(%eax)		# counter
 	je reschedule
 ret_from_sys_call:
-	movl _current,%eax		# task[0] cannot have signals
-	cmpl _task,%eax
+	movl current,%eax		# task[0] cannot have signals
+	cmpl task,%eax
 	je 3f
 	cmpw $0x0f,CS(%esp)		# was old code segment supervisor ?
 	jne 3f
@@ -116,7 +116,7 @@
 	movl %ebx,signal(%eax)
 	incl %ecx
 	pushl %ecx
-	call _do_signal
+	call do_signal
 	popl %eax
 3:	popl %eax
 	popl %ebx
@@ -127,8 +127,8 @@
 	pop %ds
 	iret
 
-.align 2
-_coprocessor_error:
+.align 4
+coprocessor_error:
 	push %ds
 	push %es
 	push %fs
@@ -142,10 +142,10 @@
 	movl $0x17,%eax
 	mov %ax,%fs
 	pushl $ret_from_sys_call
-	jmp _math_error
+	jmp math_error
 
-.align 2
-_device_not_available:
+.align 4
+device_not_available:
 	push %ds
 	push %es
 	push %fs
@@ -162,20 +162,20 @@
 	clts				# clear TS so that we can use math
 	movl %cr0,%eax
 	testl $0x4,%eax			# EM (math emulation bit)
-	je _math_state_restore
+	je math_state_restore
 	pushl %ebp
 	pushl %esi
 	pushl %edi
-	call _math_emulate
+	call math_emulate
 	popl %edi
 	popl %esi
 	popl %ebp
 	ret
 
-.align 2
-_timer_interrupt:
+.align 4
+timer_interrupt:
 	push %ds		# save ds,es and put kernel data space
-	push %es		# into them. %fs is used by _system_call
+	push %es		# into them. %fs is used by system_call
 	push %fs
 	pushl %edx		# we save %eax,%ecx,%edx as gcc doesn't
 	pushl %ecx		# save those across function calls. %ebx
@@ -186,27 +186,27 @@
 	mov %ax,%es
 	movl $0x17,%eax
 	mov %ax,%fs
-	incl _jiffies
+	incl jiffies
 	movb $0x20,%al		# EOI to interrupt controller #1
 	outb %al,$0x20
 	movl CS(%esp),%eax
 	andl $3,%eax		# %eax is CPL (0 or 3, 0=supervisor)
 	pushl %eax
-	call _do_timer		# 'do_timer(long CPL)' does everything from
+	call do_timer		# 'do_timer(long CPL)' does everything from
 	addl $4,%esp		# task switching to accounting ...
 	jmp ret_from_sys_call
 
-.align 2
-_sys_execve:
+.align 4
+sys_execve:
 	lea EIP(%esp),%eax
 	pushl %eax
-	call _do_execve
+	call do_execve
 	addl $4,%esp
 	ret
 
-.align 2
-_sys_fork:
-	call _find_empty_process
+.align 4
+sys_fork:
+	call find_empty_process
 	testl %eax,%eax
 	js 1f
 	push %gs
@@ -214,11 +214,11 @@
 	pushl %edi
 	pushl %ebp
 	pushl %eax
-	call _copy_process
+	call copy_process
 	addl $20,%esp
 1:	ret
 
-_hd_interrupt:
+hd_interrupt:
 	pushl %eax
 	pushl %ecx
 	pushl %edx
@@ -235,10 +235,10 @@
 	jmp 1f			# give port chance to breathe
 1:	jmp 1f
 1:	xorl %edx,%edx
-	xchgl _do_hd,%edx
+	xchgl do_hd,%edx
 	testl %edx,%edx
 	jne 1f
-	movl $_unexpected_hd_interrupt,%edx
+	movl $unexpected_hd_interrupt,%edx
 1:	outb %al,$0x20
 	call *%edx		# "interesting" way of handling intr.
 	pop %fs
@@ -249,7 +249,7 @@
 	popl %eax
 	iret
 
-_floppy_interrupt:
+floppy_interrupt:
 	pushl %eax
 	pushl %ecx
 	pushl %edx
@@ -264,10 +264,10 @@
 	movb $0x20,%al
 	outb %al,$0x20		# EOI to interrupt controller #1
 	xorl %eax,%eax
-	xchgl _do_floppy,%eax
+	xchgl do_floppy,%eax
 	testl %eax,%eax
 	jne 1f
-	movl $_unexpected_floppy_interrupt,%eax
+	movl $unexpected_floppy_interrupt,%eax
 1:	call *%eax		# "interesting" way of handling intr.
 	pop %fs
 	pop %es
@@ -277,7 +277,7 @@
 	popl %eax
 	iret
 
-_parallel_interrupt:
+parallel_interrupt:
 	pushl %eax
 	movb $0x20,%al
 	outb %al,$0x20
diff -ur linux-0.11/lib/Makefile linux-0.11-deb/lib/Makefile
--- linux-0.11/lib/Makefile	1991-12-02 11:16:05.000000000 +0800
+++ linux-0.11-deb/lib/Makefile	2013-11-21 16:24:08.991245134 +0800
@@ -6,13 +6,13 @@
 # unless it's something special (ie not a .c file).
 #
 
-AR	=gar
-AS	=gas
-LD	=gld
+AR	=ar
+AS	=as --32
+LD	=ld -m elf_i386
 LDFLAGS	=-s -x
 CC	=gcc
-CFLAGS	=-Wall -O -fstrength-reduce -fomit-frame-pointer -fcombine-regs \
-	-finline-functions -mstring-insns -nostdinc -I../include
+CFLAGS	=-m32 -Wall -O -fstrength-reduce -fomit-frame-pointer  \
+	-finline-functions  -nostdinc -I../include -mtune=i386
 CPP	=gcc -E -nostdinc -I../include
 
 .c.s:
diff -ur linux-0.11/lib/malloc.c linux-0.11-deb/lib/malloc.c
--- linux-0.11/lib/malloc.c	1991-12-02 11:15:20.000000000 +0800
+++ linux-0.11-deb/lib/malloc.c	2013-11-21 17:52:10.857533801 +0800
@@ -153,7 +153,8 @@
 		free_bucket_desc = bdesc->next;
 		bdesc->refcnt = 0;
 		bdesc->bucket_size = bdir->size;
-		bdesc->page = bdesc->freeptr = (void *) cp = get_free_page();
+		cp = (char*)get_free_page();
+		bdesc->page = bdesc->freeptr = (void *) cp;
 		if (!cp)
 			panic("Out of memory in kernel malloc()");
 		/* Set up the chain of free objects */
diff -ur linux-0.11/Makefile linux-0.11-deb/Makefile
--- linux-0.11/Makefile	1991-12-06 11:12:46.000000000 +0800
+++ linux-0.11-deb/Makefile	2013-11-22 17:53:48.470559578 +0800
@@ -7,12 +7,12 @@
 AS86	=as86 -0 -a
 LD86	=ld86 -0
 
-AS	=gas
-LD	=gld
+AS	=as --32
+LD	=ld -m elf_i386 -Ttext 0 -e startup_32
 LDFLAGS	=-s -x -M
 CC	=gcc $(RAMDISK)
-CFLAGS	=-Wall -O -fstrength-reduce -fomit-frame-pointer \
--fcombine-regs -mstring-insns
+CFLAGS	=-m32 -Wall -O -fstrength-reduce -fomit-frame-pointer -mtune=i386 
+ 
 CPP	=cpp -nostdinc -Iinclude
 
 #
@@ -20,7 +20,7 @@
 # This can be either FLOPPY, /dev/xxxx or empty, in which case the
 # default of /dev/hd6 is used by 'build'.
 #
-ROOT_DEV=/dev/hd6
+ROOT_DEV=FLOPPY
 
 ARCHIVES=kernel/kernel.o mm/mm.o fs/fs.o
 DRIVERS =kernel/blk_drv/blk_drv.a kernel/chr_drv/chr_drv.a
@@ -31,7 +31,7 @@
 	$(CC) $(CFLAGS) \
 	-nostdinc -Iinclude -S -o $*.s $<
 .s.o:
-	$(AS) -c -o $*.o $<
+	$(AS) -o $*.o $<
 .c.o:
 	$(CC) $(CFLAGS) \
 	-nostdinc -Iinclude -c -o $*.o $<
diff -ur linux-0.11/mm/Makefile linux-0.11-deb/mm/Makefile
--- linux-0.11/mm/Makefile	1991-12-02 11:21:45.000000000 +0800
+++ linux-0.11-deb/mm/Makefile	2013-11-21 16:24:48.271156681 +0800
@@ -1,9 +1,9 @@
 CC	=gcc
-CFLAGS	=-O -Wall -fstrength-reduce -fcombine-regs -fomit-frame-pointer \
-	-finline-functions -nostdinc -I../include
-AS	=gas
-AR	=gar
-LD	=gld
+CFLAGS	=-O -Wall -fstrength-reduce  -fomit-frame-pointer \
+	-finline-functions -nostdinc -I../include -mtune=i386 -m32
+AS	=as --32
+AR	=ar
+LD	=ld -m elf_i386
 CPP	=gcc -E -nostdinc -I../include
 
 .c.o:
diff -ur linux-0.11/mm/memory.c linux-0.11-deb/mm/memory.c
--- linux-0.11/mm/memory.c	1991-12-03 08:48:01.000000000 +0800
+++ linux-0.11-deb/mm/memory.c	2013-11-21 15:34:32.310842759 +0800
@@ -52,7 +52,7 @@
 static long HIGH_MEMORY = 0;
 
 #define copy_page(from,to) \
-__asm__("cld ; rep ; movsl"::"S" (from),"D" (to),"c" (1024):"cx","di","si")
+__asm__("cld ; rep ; movsl"::"S" (from),"D" (to),"c" (1024))
 
 static unsigned char mem_map [ PAGING_PAGES ] = {0,};
 
@@ -78,7 +78,7 @@
 	:"=a" (__res)
 	:"0" (0),"i" (LOW_MEM),"c" (PAGING_PAGES),
 	"D" (mem_map+PAGING_PAGES-1)
-	:"di","cx","dx");
+	:"dx");
 return __res;
 }
 
diff -ur linux-0.11/mm/page.s linux-0.11-deb/mm/page.s
--- linux-0.11/mm/page.s	1991-10-02 22:16:30.000000000 +0800
+++ linux-0.11-deb/mm/page.s	2013-11-22 09:51:45.064334067 +0800
@@ -9,9 +9,9 @@
  * the real work is done in mm.c
  */
 
-.globl _page_fault
+.globl page_fault
 
-_page_fault:
+page_fault:
 	xchgl %eax,(%esp)
 	pushl %ecx
 	pushl %edx
@@ -27,9 +27,9 @@
 	pushl %eax
 	testl $1,%eax
 	jne 1f
-	call _do_no_page
+	call do_no_page
 	jmp 2f
-1:	call _do_wp_page
+1:	call do_wp_page
 2:	addl $8,%esp
 	pop %fs
 	pop %es
diff -ur linux-0.11/tools/build.c linux-0.11-deb/tools/build.c
--- linux-0.11/tools/build.c	1991-12-04 21:14:34.000000000 +0800
+++ linux-0.11-deb/tools/build.c	2013-11-22 19:52:32.258325712 +0800
@@ -37,6 +37,9 @@
 #define DEFAULT_MAJOR_ROOT 3
 #define DEFAULT_MINOR_ROOT 6
 
+#define MAJOR(a) (((unsigned)(a))>>8)
+#define MINOR(a) ((a)&0xff)
+
 /* max nr of sectors of setup: don't change unless you also change
  * bootsect etc */
 #define SETUP_SECTS 4
@@ -155,7 +158,7 @@
 		die("Unable to open 'system'");
 	if (read(id,buf,GCC_HEADER) != GCC_HEADER)
 		die("Unable to read header of 'system'");
-	if (((long *) buf)[5] != 0)
+	if (((long *) buf)[6] != 0)
 		die("Non-GCC header of 'system'");
 	for (i=0 ; (c=read(id,buf,sizeof buf))>0 ; i+=c )
 		if (write(1,buf,c)!=c)
